### ISSUE TITLE
Possible to forward notifications over SSH?

### ISSUE THREAD

**Issue by tonycpsu** [marzocchi/zsh-notify#46]
I really like this plugin, but it'd be great if I could set it up on a remote host and have the notifications forwarded somehow to show up in OSX.  Is this possible?

**Comment by marzocchi** [comment]
Hello! Currently, there's no support for remote notifications now.

I've been toying with the idea of posting locally notifications generated on a ssh host, but I haven't been able to find any solution that didn't involve opening ports on the local machine, and have the remote host connect back, which is something I prefer to avoid.

**Comment by tonycpsu** [comment]
I don't know of a fully cross-platform solution, but I use iTerm2, and it supports an escape sequence:

that posts a notification.  See [here](http://www.sergeymarkov.com/blog/2013/07/using-iterm-and-growl/) for details.  Looks like you have to do it slightly differently for screen, and neither the screen nor normal way of doing it seems to work for me under tmux [see also](https://groups.google.com/forum/#!topic/iterm2-discuss/ZlTtxVIjrgg), but I do think the escape code solution has merit, and would work for screen users and those who don't use a terminal multiplexer.

**Comment by marzocchi** [comment]
Interesting! The downside is that [iTerm's escapes don't allow customization of icon and title](https://iterm2.com/documentation-escape-codes.html), so I'm somewhat hesitant in adding support for it in master.

However, zsh-notify allows you to set a custom notifier script, so you should still be able to get remote notifications with iTerm with this setup (haven't really tested it with `tmux` or `screen` yet, but it it works, I'll add it to the README.md).

You need to use `zsh` with the `zsh-notify` plugin on the remote host, then

1. create a script somewhere in your `$PATH`, eg.: `~/bin/iterm-notifier`

        #!/usr/bin/env zsh

        message=$(</dev/stdin)
        printf "\033]9;$message\a"

2. configure zsh-notify in your `.zshrc` to use the script:

        zstyle ':notify:*' notifier ~/bin/iterm-notifier

Now when a long running command ends on the server, the plugin on the server will emit the escape sequences, your iTerm will catch it and display a notification!

**Comment by marzocchi** [comment]
Uh, ignore previous comment: as I _actually_ recalled that it cannot work on remote anyway, since zsh-notify [currently assume that the terminal application is running on the same machine](https://github.com/marzocchi/zsh-notify/blob/master/notify-if-background#L218), it doesn't even try to send the notification when it's not :)

**Comment by tonycpsu** [comment]
Well, if you do change that, I found a [solution](http://unix.stackexchange.com/questions/34729/escape-sequences-in-output-of-script-called-from-ncurses-application) for getting notifications to show through tmux.  Putting it all together, this works for me in a standard xterm, tmux, and screen:

```bash
growl() {
   local msg="\033]9;${*}\007"
   case $TERM in
       screen*)
           printf "\033P${msg}\033\\" ;;
       tmux*)
           printf "\033Ptmux;\033${msg}\033\\";;
       *)
           printf ${msg} ;;
   esac
   return
}


```

**Comment by marzocchi** [comment]
thank you, I'll keep the snippet handy.

> Well, if you do change that

Eh, the trouble is that the "core" feature of this plugin is detecting whether the terminal is focused or not, to decide if a notification should be sent, which is kinda "challenging" to do when ssh is involved.

I could provide options to disable the checks on the terminal state, so that it could also be used on a remote session, but this would also cause notifications to be posted even when, say, you manually close an editor in a focused terminal window.

**Comment by tonycpsu** [comment]
Well, [this](http://unix.stackexchange.com/a/9607/216766) StackExchange post shows a pretty reliable way to detect whether the shell is controlled by an SSH session.  At the worst, your plugin could probably check that, so that the focused check could be ignored only for SSH sessions, but retained when running locally.

Looking at how you're detecting focus with applescript and such, I agree that might be a problem.  However, it looks from [this vim plugin])https://github.com/sjl/vitality.vim/blob/master/plugin/vitality.vim) as if iTerm2 supports some escape codes for reporting when a window gains / loses focus.  This would probably be a better way of doing the check with iTerm2 anyway, since it doesn't involve launching an applescript -- just some escape codes to send and respond to.

Anyway, I get the sense that maybe this request is getting larger, so if this isn't something you want to tackle, maybe I can give it a go at some point.

**Comment by marzocchi** [comment]
> Anyway, I get the sense that maybe this request is getting larger, so if this isn't something you want to tackle

actually, not having notifications on ssh is a pain point for me (it's only mitigated by the fact that I can't have zsh on the servers I use at work :D)

i'd be very happy to trash applescript/xdotool support if I can find a different solution, and I was not aware that [terminals can report focus with escapes](https://github.com/sjl/vitality.vim/blob/master/plugin/vitality.vim#L66-L67).

still have to figure out how to "capture" escapes sent from the terminal from a script!

**Comment by tonycpsu** [comment]
> still have to figure out how to "capture" escapes sent from the terminal from a script!

That should just be a matter of `bindkey ESCAPE_SEQUENCE`, right?  I just tried this and it seemed to work:

```bash
printf "\033[?1004h"
bindkey -s "\033[O" "echo lost focus^M"
```

**Comment by marzocchi** [comment]
Hello again, sorry for the delay.

I expanded on your suggestion, trying to [use widgets to store the focus state in a variable](https://gist.github.com/marzocchi/a2993ee75a4be887cbf6b31276786e35) (plus some logging). I used also a `precmd` hook because that's where I need to know the focus state in `zsh-notify`.

**tl;dr:** seems like it can work, but with a different hook than `precmd`, as it happens too early.

I tried with some scenarios:

1. sleep for a while, make terminal gain and lose the focus some times (causing a bunch of `^[[O` and `^[[I` to pile up on the screen), and let the command finish with an unfocused terminal. This is what ended up in the log:

        showing prompt; active? 1
        lost focus
        got focus
        lost focus
        got focus
        lost focus
        got focus
        lost focus
        got focus
        lost focus

    Zsh first runs the `precmd` hook, then processes the escape sequences that where generated by (un)focusing the terminal during `sleep`. 

2. sleep and (un)focus a bunch of time again, but instead of waiting, focus the terminal and hit `^C`. This time, the log shows:

        showing prompt; active? 1

    The `got-focus`/`lost-focus` widgets were not run at all, even if as before I had a bunch of `^[[O` and `^[[I` piling up on the screen.

3. again sleep, (un)focus a bunch of time, and kill the program from another shell. Log shows:

        showing prompt; active? 1
        lost focus
        got focus
        lost focus
        got focus
        lost focus
        got focus
        lost focus
    

To sum up:

- the escape sequences can reach the line editor only after `precmd` (eg.: after prompt is shown), so a different hook must be used;

- in some cases the escape sequences might not reach the line editor

  - a case like nr. 2 seems to be ok-ish: the terminal was active and a notification was not needed;

  - there might be some corner cases in which a notification might be needed, but can't be sent: a program that consumes escape sequences on its own, but terminates without user interaction;

**Comment by tonycpsu** [comment]
Interesting findings.  So it seems like the `preexec` hook would be the way to go?  Have you tried using that instead of `precmd`?

**Comment by marzocchi** [comment]
as far as my plugin is concerned, `preexec` is “when you press enter to start a command”, while `precmd` is “when the program is finished”, so I can't evaluate the focus state at `preexec` because it's too early.

while I'm not particularly “attached” to the AppleScript based solution I'm currently using, an escapes-based solution (if possible) would require a considerable amount of work and only get me as far as being able to present text only notifications and only iTerm2 – [so I would not have doge bark at me when I mistype something](https://gist.github.com/marzocchi/14c47a49643389029a2026b4d4fec7ae), which is kinda my favourite feature :D

**Comment by tonycpsu** [comment]
Fair enough -- thanks for looking into it.
